### PR TITLE
8338924: C1: assert(0 <= i && i < _len) failed: illegal index 5 for length 5

### PR DESCRIPTION
**Pull request by matias9927**
The change in [JDK-8335664](https://bugs.openjdk.org/browse/JDK-8335664) caused a crash when initializing basic blocks with `-Xcomp`. This change introduces a check to see if JSR is the last bytecode in its method so that expected behavior matches the previous patch. Verified with tier 1-6 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338924](https://bugs.openjdk.org/browse/JDK-8338924): C1: assert(0 &lt;= i &amp;&amp; i &lt; _len) failed: illegal index 5 for length 5 (**Bug** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Contributors
 * Dean Long `<dlong@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20732/head:pull/20732` \
`$ git checkout pull/20732`

Update a local copy of the PR: \
`$ git checkout pull/20732` \
`$ git pull https://git.openjdk.org/jdk.git pull/20732/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20732`

View PR using the GUI difftool: \
`$ git pr show -t 20732`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20732.diff">https://git.openjdk.org/jdk/pull/20732.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20732#issuecomment-2318428527)

### CODE REVIEW

**Comment by dean-long** [comment]
Please add a loop to the test so it triggers this issue without -Xcomp.

File: test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java [hunk 1]
```diff
@@ -24,17 +24,22 @@
...
+ * @run main/othervm
+ *      -Xbatch
+ *      LastJsrTest
  */
 
 public class LastJsrTest {
     public static void main(String[] args) {
-        LastJsr.test();
-        LastJsrReachable.test();
+        for (int i = 0; i < 1000; ++i) {
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java:39]: Don't you need 10,000 in your loop to trigger compilation?

**Review by coleenp — COMMENTED** [review]
Just a comment.

File: test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java (hunk 1 above)
Comment (dean-long) [reply test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java:39]: Yes for C2, but this is enough for C1, the only compiler that needs this fix.  I wanted to make sure C1 compilation was triggered by default without -Xcomp.  Testing tiers that use -Xcomp will make sure it passes with C2.

File: test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java [hunk 2]
```diff
@@ -24,17 +24,22 @@
...
  * @bug 8335664
+ * @bug 8338924
  * @summary Ensure a program that ends with a JSR does not crash
  * @library /test/lib
  * @compile LastJsr.jasm
  * @compile LastJsrReachable.jasm
- * @run main/othervm LastJsrTest
+ * @run main/othervm
+ *      -Xbatch
+ *      LastJsrTest
```
Comment (vnkozlov) [inline test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java:34]: I suggest to keep old `command` too.
And I don't like spliting new command into several lines when it is very short.

File: src/hotspot/share/compiler/methodLiveness.cpp [hunk 3]
```diff
@@ -222,6 +222,9 @@ void MethodLiveness::init_basic_blocks() {
           dest = _block_map->at(bytes.get_dest());
           assert(dest != nullptr, "branch destination must start a block.");
           dest->add_normal_predecessor(current_block);
+
+          if (bci + Bytecodes::length_for(code) >= method_len) break;
+
```
Comment (vnkozlov) [inline src/hotspot/share/compiler/methodLiveness.cpp:227]: Please, use code style which use `{}` and put body on separate line as you did in `c1_GraphBuilder.cpp`
You will not need empty lines around it then.

File: src/hotspot/share/compiler/methodLiveness.cpp [hunk 4]
```diff
@@ -232,6 +235,9 @@ void MethodLiveness::init_basic_blocks() {
           dest = _block_map->at(bytes.get_far_dest());
           assert(dest != nullptr, "branch destination must start a block.");
           dest->add_normal_predecessor(current_block);
+
+          if (bci + Bytecodes::length_for(code) >= method_len) break;
+
```
Comment (vnkozlov) [inline src/hotspot/share/compiler/methodLiveness.cpp:240]: Same here.

File: test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java (hunk 1 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/interpreter/LastJsrTest.java:39]: Ok.

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/c1/c1_GraphBuilder.cpp [hunk 5]
```diff
@@ -1389,6 +1389,11 @@ void GraphBuilder::jsr(int dest) {
   // If the bytecodes are strange (jumping out of a jsr block) then we
   // might end up trying to re-parse a block containing a jsr which
   // has already been activated. Watch for this case and bail out.
+  if (next_bci() >= method()->code_size()) {
+      // This can happen if the subroutine does not terminate with a ret,
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_GraphBuilder.cpp:1393]: Indentation is incorrect here (should be 2 whitespace instead of 4).

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

**Review by vnkozlov — APPROVED** [review]
Update is good.

**Review by TobiHartmann — APPROVED** [review]
Still good.

**Comment by matias9927** [comment]
Thanks for the reviews @TobiHartmann and @vnkozlov! Also thank you for the assistance @dean-long!
/integrate

**Pull request closed without merging**